### DISCUSSION TITLE
missleading error message regarding argument-dependent return type of a generic function with default parameter value

### DISCUSSION THREAD

**Discussion by Sorontik** [microsoft/pyright#5405] [Q&A] [answered]
First off, I'm no python type-system expert, so please tell me, if I use terms incorrectly or understand something wrong.

I'm using an up-to-date VS Code with up-to-date normal release python extensions and type-checking set to 'basic'.

I just spent quite some time disecting a pyright error in a larger project, because in my opinion, the error is highly counter-intuitive and missleading. 
I was able to boil it down to the reduced minimal example below and i know what the root cause is, but i don't understand the error message at all.

_(My actual questions are stated at the end)_

In the example:
```python
from typing import Type, TypeVar


class A:
    pass

class B(A):
    pass


o : A = B()


T = TypeVar("T", bound=A)

def factory(elc : Type[T] = A) -> list[T]:
    return []


l1 : list[A] = factory()
l2 : list[B] = factory(B)

l3 : list[B] = factory()  # ! error message 1 here


def factory2() -> list[A]:
    return []

l4 : list[B] = factory2() # ! error message 2 here
```
pyright correctly reports an error at the assignment of `l3`.
The problem is the error message (error message 1), which is:
```
Argument of type "Type[A]" cannot be assigned to parameter "elc" of type "Type[T@factory]" in function "factory"
  Type "B" cannot be assigned to type "A"    Pylance[reportGeneralTypeIssues]
```

Which does not make sense to me, since `T` is bound to `A`, `Type[A]` should be 'assignable'  to `Type[T]` and even though i'm not sure about the semantics of the types themselves here, an object of type `B` can be assigned to a variable of type `A` perfectly well, as the assignment of `o` demonstrates.

At first i thought that pyright looks at this backwards:
1.  the typ of `l3` is `list[B]`
2. therefore the return type of the call must be `list[B]`
3. therefore `T` must be `B`
4. therefore, the default value for `elc` = `A` has the wrong type

this, however, does not explain the second line of the error message, which complains that `B` cannot be assigned to `A`, so with the two types swapped.

Intuitively, i would approach it like this (and naively expect pyright to do it in a similar way):
1. there is no argument provided, so `elc` has the default value `A`
2. therefore `T` 'is' `A`
3. therefore the return type is `list[A]`

which makes me expect an error message similar to the error message 2, that appears if the generics are removed and the return type is stated explicitly:
```
Expression of type "list[A]" cannot be assigned to declared type "list[B]"
  "list[A]" is incompatible with "list[B]"
    TypeVar "_T@list" is invariant
      "A" is incompatible with "B"
```

Since, from a high-level point of view, both errors are just flavors of the same problem, I would expect them to produce similar error messages (though the first one obviously should still mention the TypeVar), but that might be 

**So this brings me to my Questions:**
1. Please explain to me the reasoning behind both lines of the first error message
2. and why it is so vastly different from the second one even though in both cases the core problem is the same

**Comment by erictraut** [comment, marked as answer]
There was a bug that resulted in the two types being swapped in the error message. It's not surprising that you found it confusing. This will be addressed in the next release. Thanks for reporting the issue.
